core/iwasm/interpreter/wasm_loader.c

```diff
@@ -12105,6 +12111,62 @@ wasm_loader_prepare_bytecode(WASMModule *module, WASMFunction *func,
                                        error_buf_size)) {
                     goto fail;
                 }
+                tbl_elem_type =
+                    table_idx < module->import_table_count
+                        ? module->import_tables[table_idx]
+                              .u.table.table_type.elem_type
+                        : module->tables[table_idx - module->import_table_count]
+                              .table_type.elem_type;
+
+#if WASM_ENABLE_GC == 0 && WASM_ENABLE_REF_TYPES != 0
+                if (tbl_elem_type != VALUE_TYPE_FUNCREF) {
+                    set_error_buf_v(error_buf, error_buf_size,
+                                    "type mismatch: instruction requires table "
+                                    "of functions but table %u has externref",
+                                    table_idx);
+                    goto fail;
+                }
+#elif WASM_ENABLE_GC != 0
+                /* For call_indirect, if table element is heaptype, then it's
+                 * valid only if it's (ref null? func) or (ref null? $t) where t
+                 * is a concrete heap type that match funcref */
```

Comment:
The below checks make a little confusing, from the GC spec
https://webassembly.github.io/gc/core/valid/instructions.html#xref-syntax-instructions-syntax-instr-control-mathsf-call-indirect-x-y

https://webassembly.github.io/gc/core/valid/matching.html#match-heaptype

Per my understanding, it means that the table element's type is subtype of funcref, the code seems can be just like:
```C
    elem_ref_type = ...;
    if (!wasm_reftype_is_subtype_of(
                                    elem_type, elem_ref_type, VALUE_TYPE_FUNCREF, NULL,
                                    module->types, module->type_count)) {
        set_error_buf(...);
        goto fail;
    }
```